File: Python/DFS-BFS/5-4.py
'''
count the shortest steps to get to (n,m) from (1,1)
the only path you can take is 1's

EX) 5x6 map
101010
111111
000001
111111
111111 -----------> 10
'''
from collections import deque

n, m = map(int, input().split())

graph = []
for i in range(n):
    graph.append(list(map(int,input())))

#directions possible
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

#bfs
def bfs(x,y):
    #init queue    
    q = deque()
    q.append((x,y))

    while q:
        x,y = q.popleft()
        #check 4 directions
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]

            #if out of bounds ignore
            if nx < 0 or ny < 0 or nx >= n or ny >= m:
                continue
            #if move not possible ignore
            if graph[nx][ny] == 0:
                continue
            #if move possible append to queue
            if graph[nx][ny] == 1:
                graph[nx][ny] = graph[x][y] + 1
                q.append((nx,ny))

    return graph[n-1][m-1]

print(bfs(0,0))

File: C++/DFS-BFS/5-4.cpp
/*
count the shortest steps to get to (n,m) from (1,1)
the only path you can take is 1's

EX) 5x6 map
101010
111111
000001
111111
111111 -----------> 10
*/
#include <bits/stdc++.h>

using namespace std;

int n, m;
int graph[1000][1000];

//directions possible
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

//bfs
int bfs(int x, int y){
    //init queue
    queue<pair<int,int>> q;
    q.push({x,y});

    while (!q.empty()){
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        //check 4 directions
        for (int i = 0; i < 4; i++){
            int nx = x+dx[i];
            int ny = y+dy[i];

            //if out of bounds ignore
            if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
            //if move not possible ignore
            if (graph[nx][ny] == 0) continue;
            //if move possible append to queue
            if (graph[nx][ny] == 1){
                graph[nx][ny] = graph[x][y] + 1;
                q.push({nx,ny});
            }
        }
    }
    return graph[n-1][m-1];
}

int main(){
    cin >> n >> m;

    for (int i = 0; i < n; i++){
        for (int j = 0; j < m; j++){
            scanf("%1d", &graph[i][j]);
        }
    }

    cout << bfs(0,0) << '\n';
}